File: python/PhoneNumber.py
class PhoneNumber:
    def __init__(self, number):
        clear_number = ''
        for letter in number:
            if letter.isdigit():
                clear_number += letter
            elif letter.isalpha(): # if a phone number has letters in place of some digits.
                raise ValueError("letters not permitted")
            elif letter in ' ()-.+':
                pass
            else:  # if a phone number has punctuation in place of some digits.
                raise ValueError("punctuations not permitted")
        # if a phone number has less than 10 digits.
        if len(clear_number) < 10:
            raise ValueError("must not be fewer than 10 digits")
        # if a phone number has more than 11 digits.
        if len(clear_number) > 11:
            raise ValueError("must not be greater than 11 digits")
        # if a phone number has 11 digits, but starts with a number other than 1.
        if len(clear_number) == 11 and clear_number[0] != '1':
            raise ValueError("11 digits must start with 1")
        # if a phone number has an exchange code that starts with 0.
        if clear_number[-7] == '0':
            raise ValueError("exchange code cannot start with zero")
        # if a phone number has an exchange code that starts with 1.
        if clear_number[-7] == '1':
            raise ValueError("exchange code cannot start with one")
        # if a phone number has an area code that starts with 0.
        if clear_number[-10] == '0':
            raise ValueError("area code cannot start with zero")
        # if a phone number has an area code that starts with 1.
        if clear_number[-10] == '1':
            raise ValueError("area code cannot start with one")
        self.number = clear_number[-10:]
        self.area_code = self.number[:3]

    def pretty(self):
        return f'({self.number[:3]})-{self.number[3:6]}-{self.number[6:]}'

File: javascript/PhoneNumber.js
//
// This is only a SKELETON file for the 'Phone Number' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const clean = (input) => {
  if(/[a-z]/gi.test(input)){throw new Error('Letters not permitted') }
  if(/([\072-\100])/gi.test(input)){throw new Error('Punctuations not permitted') }
  if(input.match(/\d/gi).length < 10){ throw new Error('Must not be fewer than 10 digits') }
  if(input.match(/\d/gi).length > 11){ throw new Error('Must not be greater than 11 digits') }
  let toReturn = input.match(/\d/gi)
  if(toReturn.length == 11){
    if(toReturn[0] != '1'){throw Error('11 digits must start with 1')}
    toReturn = toReturn.slice(1)
    if(toReturn[0] == 0){throw Error('Area code cannot start with zero')}
    if(toReturn[0] == 1){throw Error('Area code cannot start with one')}
    if(toReturn[3] == 0){throw Error('Exchange code cannot start with zero')}
    if(toReturn[3] == 1){throw Error('Exchange code cannot start with one')}
    return toReturn.join('')
  }
  if(toReturn.length == 10){
    if(toReturn[0] == 0){throw Error('Area code cannot start with zero')}
    if(toReturn[0] == 1){throw Error('Area code cannot start with one')}
    if(toReturn[3] == 0){throw Error('Exchange code cannot start with zero')}
    if(toReturn[3] == 1){throw Error('Exchange code cannot start with one')}
    return toReturn.join('')
  }
};

